src/target/riscv/riscv-013.c

```diff
@@ -3396,310 +3396,524 @@ static int write_memory_abstract(struct target *target, target_addr_t address,
+ * - Appropriate program loaded to program buffer.
+ * - DM_ABSTRACTAUTO_AUTOEXECDATA is set.
+ */
+static int read_memory_progbuf_inner_run_and_process_batch(struct target *target,
+		struct riscv_batch *batch, struct memory_access_info access,
+		uint32_t start_index, uint32_t elements_to_read, uint32_t *elements_read)
+{
+	RISCV013_INFO(info);
 
-				dmi_write(target, DM_ABSTRACTAUTO, 0);
+	if (batch_run(target, batch) != ERROR_OK)
+		return ERROR_FAIL;
 
-				uint32_t dmi_data0, dmi_data1 = 0;
-				/* This is definitely a good version of the value that we
-				 * attempted to read when we discovered that the target was
-				 * busy. */
-				if (dmi_read(target, &dmi_data0, DM_DATA0) != ERROR_OK) {
-					riscv_batch_free(batch);
-					goto error;
-				}
-				if (size > 4 && dmi_read(target, &dmi_data1, DM_DATA1) != ERROR_OK) {
-					riscv_batch_free(batch);
-					goto error;
-				}
+	uint32_t abstractcs;
 
-				/* See how far we got, clobbering dmi_data0. */
-				if (increment == 0) {
-					uint64_t counter;
-					result = register_read_direct(target, &counter, GDB_REGNO_A0);
-					next_index = counter;
-				} else {
-					uint64_t next_read_addr;
-					result = register_read_direct(target, &next_read_addr,
-												  GDB_REGNO_S0);
-					next_index = (next_read_addr - address) / increment;
-				}
-				if (result != ERROR_OK) {
-					riscv_batch_free(batch);
-					goto error;
-				}
+	if (wait_for_idle(target, &abstractcs) != ERROR_OK)
+		return ERROR_FAIL;
 
-				uint64_t value64 = (((uint64_t)dmi_data1) << 32) | dmi_data0;
-				buf_set_u64(buffer + (next_index - 2) * size, 0, 8 * size, value64);
-				log_memory_access64(address + (next_index - 2) * size, value64, size, true);
+	uint32_t elements_to_fetch_from_batch;
 
-				/* Restore the command, and execute it.
-				 * Now DM_DATA0 contains the next value just as it would if no
-				 * error had occurred. */
-				dmi_write_exec(target, DM_COMMAND, command, true);
-				next_index++;
+	info->cmderr = get_field(abstractcs, DM_ABSTRACTCS_CMDERR);
+	switch (info->cmderr) {
+	case CMDERR_NONE:
+		LOG_TARGET_DEBUG(target, "successful (partial?) memory read [%"
+				PRIu32 ", %" PRIu32 ")", start_index, start_index + elements_to_read);
+		elements_to_fetch_from_batch = elements_to_read;
+		break;
+	case CMDERR_BUSY:
+		LOG_TARGET_DEBUG(target, "memory read resulted in busy response");
+		if (read_memory_progbuf_inner_on_ac_busy(target, start_index,
+					elements_to_read, &elements_to_fetch_from_batch, access)
+				!= ERROR_OK)
+			return ERROR_FAIL;
+		break;
+	default:
+		LOG_TARGET_DEBUG(target, "error when reading memory, cmderr=0x%" PRIx32,
+				info->cmderr);
+		riscv013_clear_abstract_error(target);
+		return ERROR_FAIL;
+	}
 
-				dmi_write(target, DM_ABSTRACTAUTO,
-						1 << DM_ABSTRACTAUTO_AUTOEXECDATA_OFFSET);
+	if (read_memory_progbuf_inner_extract_batch_data(target, batch, start_index,
+				elements_to_fetch_from_batch, elements_read, access) != ERROR_OK)
```

Comment:
Detail: Since the function is called "extract" batch data, you can also use the word "extract" instead of "fetch" in the variable name - to make it consistent.

```suggestion
				elements_to_extract_from_batch, elements_read, access) != ERROR_OK)
```

Reply:
Addressed.